hw07.c

```diff
@@ -0,0 +1,246 @@
+
+typedef struct {
+    byte blue;
+    byte green;
+    byte red;
+    byte alpha;
+} RGBA;
+
+typedef struct{
+    TGAheader header;
+    RGBA *data;
+}TGAImage;
+
+void watch_draw_time(TGAImage* self, const int hours, const int minutes);
+void watch_draw_time_color(TGAImage* self, const int hours, const int minutes, const RGBA* fg_color, const RGBA* bg_color);
+
+//zjistovani/kopirovani hodnoty z headru (z bytu na cislo) a predavani cisla, takze kdykoli si necham zavolat tuhle fci budu li chtit hodnotu, pac kde hodnote se dostanu
+//ale je to identifikovano jako unsigned char, proste byte, musi memcpy do integer promene vytvorene (co prima konverze typova??? , hraje roli to ze to je pole v nekterych pripadech??)
+int tgaheader_width( const TGAheader * self ) {
+    int width = 0;
+
+    memcpy( &width, self->width, 2 );
+    printf("width: %d\n",width);
+
+    return width;
+}
+
+int tgaheader_height( const TGAheader * self ) {
+    int height = 0;
+
+    memcpy( &height, self->height, 2 );
+    printf("height: %d\n",height);
+
+    return height;
+
+}
+
+int tgaheader_channels( const TGAheader * self ) {
+    return self->depth / 8;
+}
+
+//zjistovani hodnot z headru pres fci - asi prehlednejsi
+int tga_width( const TGAImage * self ) {
+    return tgaheader_width( &self->header );
+}
+
+int tga_height( const TGAImage * self ) {
+    return tgaheader_height( &self->header );
+}
+
+int tga_channels( const TGAImage * self ) {
+    return  tgaheader_channels( &self->header );
+}
+
+
+//prevodova fce z 2d souradnic do indexu 1D odpovidajiciho serioveho elementu
+int to1D(int row, int col, int cols){
+    return col + row * cols;
+}
+
+
+//cokoliv zavola fci se souradnicema x,y bude prevedeno z 2d pole do odpovidajiciho indexu 1d pole jiz nainacializovaneho v *data
+//prehodim x, y kdyz pocatek je 0,0 v dolnim levem rohy pokud prehodim x z y tak he to jakobych ho otozil o 90/ prohodil sloupce/radky
+void set_pixel(TGAImage *self, int y, int x, RGBA *pix){
+    self->data[to1D(x,y, tga_height(self))].alpha = pix->alpha;
+    self->data[to1D(x,y, tga_height(self))].blue = pix->blue;
+    self->data[to1D(x,y, tga_height(self))].green = pix->green;
+    self->data[to1D(x,y, tga_height(self))].red = pix->red;
+}
+
+void draw_bg(TGAImage *self, RGBA *bg){
+    for (int i = 0; i < tga_height(self); i++)
+    {
+        for (int j = 0; j < tga_width(self); j++)
+        {
+            set_pixel(self, i ,j , bg);
+            printf("bg: i:%d j:%d\n",i ,j);
+        }
+
+    }
+
+}
+
+//pocita indexy nalezejici vykresleni znaku  ty posila do fce ktere zapisuje ty hodnotu do pole clena strukty *data
+void draw_one(TGAImage *self, int dx, int dy, RGBA *pix){
+    for (int i = 0 + dx; i < 195 + dx; i++)
+    {
+        for (int j = 0 + dx; j < 169 + dy; j++)
+        {
+            if (i > 15 + dx && j > 139 + dy)
+            {
+                set_pixel(self, i, j, pix);
+                printf("one: i:%d j:%d\n",i ,j);
+            }
+        }
+    }
+}
```

Comment:
Namisto tohoto aby jsi mel v kazde `draw_` funkci fory a ify, tak muzes mit jen jednu `draw` funkci, ktere pridas 4 promenne - rozsahy pro tvoje `i` a `j`, protoze to je jedinne co ve vsech tech funkcich delas.
(Nebo muzes predavat dve pole se 2 promennyma, nebo predavat jedno pole se 4 souradnicema... to uz je uplne jedno.)

A potom by jsi kreslil pro one jeden obdelnik, pro dalsi pak vice obdelniku, napr.:

```suggestion
void draw(TGAImage *self, int x_start, int x_end, y_start, y_end, RGBA *pix){
    for (int i = x_start; i < x_end; i++)
    {
        for (int j = y_start; j < y_end; j++)
        {
                set_pixel(self, i, j, pix);
                printf("one: i:%d j:%d\n",i ,j);
        }
    }
}


void draw_one(TGAImage *self, int dx, int dy, RGBA *pix){
   draw (self, dx + 15, dy + 195, dx + 139, dy + 169, pix);
}
```

`dy` a `dx` (offsety) jdou samozdrejme zakomponovat na libovolne urovni, ja to zvolil pro jednoduchost funkce `draw` tu (jako illustrativni priklad), ale samozdrejme se to da taky predavat dale funkci `draw`, cimz se tyto funkce kompletne zjednodusi (a nebudou uz vubec potreba).

Reply:
to nevim jestli ted pobiram, sem dost nevyspalej a draw bych resil az mi to aspon vytvori nejaky soubor, ktery odpovida, nevykresli mi to ani pozadi spravne, ani jeden pixel, myslel sem, neco blbe s tou hlavickou, akorat se moc nevyznam v tech hodnotach v debugru ve VS. jestli tam sou blosti nebo ne, kazdopadne sem myslel, ze vsechno memsetem nuluju a pak dodam konkretni nenulove hodnoty, ale uz jsem z toho domotanej
   memset(&tga->header, 0, sizeof(TGAheader)); //vynulovani
    tga->header.image_type = 2;
    tga->header.depth = 32;
    
    //00101000 = 40
    tga->header.descriptor |= 1 << 3; 
    tga->header.descriptor |= 1 << 5;  

    //zkopirovani hodnot sirky vysky na dva bajty
    memcpy(&tga->header.width, &width, 2 );
    memcpy(&tga->header.height, &height, 2 );


pri kompilaci ve VS mi to zadny chyby nahaze asi, ale podle me tam nemam spravny parametry pro kompilaci, v QTcreatoru mi to zas neprelozi vubec a spadne to na fault segmentation.  V terminale s gcc -o program  hw07.c -g -Wall -Wextra -pedantic -std=c99 to projde.